packages/flutter_tools/lib/src/asset.dart

```diff
@@ -0,0 +1,420 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:json_schema/json_schema.dart';
+import 'package:path/path.dart' as path;
+import 'package:yaml/yaml.dart';
+
+import 'cache.dart';
+import 'dart/package_map.dart';
+import 'globals.dart';
+
+/// An entry in an asset bundle.
+class AssetBundleEntry {
+  /// An entry backed by a File.
+  AssetBundleEntry.fromFile(this.archivePath, this.file)
+      : _contents = null;
+
+  /// An entry backed by a String.
+  AssetBundleEntry.fromString(this.archivePath, this._contents)
+      : file = null;
+
+  /// The path within the bundle.
+  final String archivePath;
+
+  /// The payload.
+  List<int> contentsAsBytes() {
+    if (_contents != null) {
+      return UTF8.encode(_contents);
+    } else {
+      return file.readAsBytesSync();
+    }
+  }
+
+  bool get isStringEntry => _contents != null;
+
+  final File file;
+  final String _contents;
+}
+
+/// A bundle of assets.
+class AssetBundle {
+  final Set<AssetBundleEntry> entries = new Set<AssetBundleEntry>();
+
+  static const String defaultManifestPath = 'flutter.yaml';
+  static const String defaultWorkingDirPath = 'build/flx';
+  static const String _kFontSetMaterial = 'material';
+  static const String _kFontSetRoboto = 'roboto';
+
+  AssetBundle();
+
+  Future<int> build({String manifestPath: defaultManifestPath,
+                     String workingDirPath: defaultWorkingDirPath,
+                     bool includeRobotoFonts: true}) async {
+    Object manifest = _loadFlutterYamlManifest(manifestPath);
+    if (manifest != null) {
+     int result = await _validateFlutterYamlManifest(manifest);
+     if (result != 0)
+       return result;
+    }
+    Map<String, dynamic> manifestDescriptor = manifest;
```

Comment:
Is this cast safe?  Maybe consider an is check first in case it's malformed.  Or maybe that's too defensive?